### ISSUE TITLE
Queue consumer EmicoRobinHq stacks up processes

### ISSUE THREAD

**Issue by arnoudhgz** [EmicoEcommerce/Magento2RobinHq#8]
At two of our clients we have this module installed and the consumer is being triggered every minute (according to the cron_schedule table). 
But we see that multiple consumers for EmicoRobinHq keep running until the server has a memory outage. 

Maybe related to this, we see that RabitMQ processes a message when the order gets the processing status. But it never appears as a 'win' in Robin.

We use:
* Module version 1.0.4
* Magento 2.3.3
* php7.2
* RabitMQ enabled and connection established

**Comment by rofokken** [comment]
Hello, 

Thank you for your question. The starting and killing of consumer processes is done by Magento. "module-message-queue" a process is started with only a --max-messages parameter. This is by default Magento "a long living process" There is also a locking mechanism in place see "vendor/magento/module-message-queue/Model/Cron/ConsumersRunner.php:143" which should prevent starting more of the same consumers.

Our crontab only facilitates a fallback when rabbitmq is not installed. See src/Cron/FileQueueConsumer.php:32. This is not what you mean right?

Could you reproduce this with a clean install? Could you provide more steps to reproduce?

Regards,

Robert

**Comment by arnoudhgz** [comment]
@rofokken I'm currently debugging it further. I now have rabitmq also locally configured (although no success yet with getting messages in the RabitMQ monitor). 

In the monitor I can see a consumer is running.
![image](https://user-images.githubusercontent.com/10864061/73273019-987e0a00-41e3-11ea-8237-56ca36440a42.png)
 
I can't provide steps to reproduce on the issue with the multiple consumers, because I don't know how to reproduce it. I just see a lot of processes with this consumer when I log in on the server.

**Comment by arnoudhgz** [comment]
@rofokken I enabled tracing on RabitMQ and now I got this message, is this correct and what you expect? I could not find where it would be transformed to an object like Robin expects 

```
================================================================================
2020-01-28 16:19:31:267: Message published

Node:         rabbit@xxxxx-ThinkPad-T470s
Connection:   127.0.0.1:41926 -> 127.0.0.1:5672
Virtual host: /
User:         guest
Channel:      1
Exchange:     magento
Routing keys: [<<"emico.robinhq">>]
Routed queues: []
Properties:   [{<<"message_id">>,longstr,
                <<"ed3ae5cd6be2b516a24b178046120476">>},
               {<<"delivery_mode">>,signedint,2}]
Payload: 
"O:33:\"Emico\\RobinHqLib\\Event\\OrderEvent\":1:{s:8:\"\u0000*\u0000order\";O:28:\"Emico\\RobinHqLib\\Model\\Order\":13:{s:14:\"\u0000*\u0000orderNumber\";s:14:\"XXXX2000059524\";s:15:\"\u0000*\u0000emailAddress\";s:22:\"xxxx@email.org\";s:7:\"\u0000*\u0000name\";s:9:\"Test Test\";s:10:\"\u0000*\u0000revenue\";d:154;s:13:\"\u0000*\u0000oldRevenue\";d:154;s:9:\"\u0000*\u0000profit\";N;s:12:\"\u0000*\u0000oldProfit\";N;s:12:\"\u0000*\u0000orderDate\";O:17:\"DateTimeImmutable\":3:{s:4:\"date\";s:26:\"2020-01-28 16:18:56.000000\";s:13:\"timezone_type\";i:3;s:8:\"timezone\";s:3:\"UTC\";}s:13:\"\u0000*\u0000firstOrder\";b:0;s:6:\"\u0000*\u0000url\";s:73:\"https:\/\/www.domain.nl\/admin\/sales\/order\/view\/order_id\/10\/\";s:14:\"\u0000*\u0000webstoreUrl\";s:38:\"https:\/\/www.domain.nl\/\";s:11:\"\u0000*\u0000listView\";N;s:15:\"\u0000*\u0000detailsViews\";N;}}"
```

**Comment by arnoudhgz** [comment]
@rofokken any update about this? Currently the main issue is that results do not appear on the Robin dashboard.

I'm also in contact with Robin and they pointed me to the case that in the screenshot **emico.robinhq** at the queue binding is mentioned (as exchange?), but in the published message **magento** is mentioned as exchange.

**Comment by bramstroker** [comment]
@arnoudhgz 

About the multiple consumers. This is very strange, and should be managed by Magento to have only one consumer at the same time running. You could have a look into the cronjob manager for the `consumer_runner` job, is this running multiple times?
Do you also see the Magento consumers (for example `bin/magento queue:consumers:start async.operations.all`) running more than one time in your process list? Do you see any patterns in the start time of the robin consumer on your server when checking with `ps -aux`, are they started each minute for example?

I would suggest the following to circumvent the issue with multiple consumer instances.
For most our enterprise shops we are not running the RabbitMQ consumers by cron but configured the consumers in supervisord (http://supervisord.org/). This way you have a seperate script per worker/consumer and things are a lot more scalable that way.
You have to add some configuration to `app/env.php`:
```
 'cron_consumers_runner' => [
        'cron_run' => false,
        'max_messages' => 50,
        'consumers' => [
        ]
```
Make sure you exclude the robinHQ consumer in the `consumers` list, but add the default magento core consumers (or better yet, also manage those with supervisor).

Next you can add the robinHQ consumer as a seperate supervisor program.
Create a `consumer-robinhq.conf` file in your `supervisor.d` directory with something like this:
```
[program:jobqueue-consumer-EmicoRobinHq]
command=bash -c 'timeout 600 php72 bin/magento queue:consumers:start EmicoRobinHq --max-messages=100'
numprocs=1
autorestart=true
autostart=true
```

Now supervisor should only start one consumer instance. After 100 messages the consumer is killed and automatically restarted by supervisor. 

The message you got while tracing looks fine.
To answer your question about the transformation for the order-post object I will give you some pointers.
The message should be handled first by the AMQP consumer:
`\Emico\RobinHq\Queue\AmqpConsumer::processMessage()`
Next the `EventProcessingService` will select a processor to handle the message.
For your message the processor will be:
`\Emico\RobinHqLib\EventProcessor\OrderEventProcessor::processEvent()`.
This will call the method on the REST client to post the payload to the dynamic order post endpoint.

I am not sure about the exchange, this should be `emico.robinhq`. We have to do some further investigation on this, but that would be next week.

Hope this helps.

**Comment by arnoudhgz** [comment]
@bramstroker the multiple processes seems a Magento bug. We solved it by changing the lock to file again instead of database.

**Comment by bramstroker** [comment]
We found a misconfiguration which caused the messages to be published to the wrong exchange. This is fixed by 3c042cb.
This should also fix your issues with the POST requests not being issued
